python/ray/data/_internal/stats.py

```diff
@@ -275,137 +332,302 @@ def summary_string(
+        if self.extra_metrics:
+            indent = "\t" if stage_stats_summary.is_substage else ""
+            out += indent
+            out += "* Extra metrics: " + str(self.extra_metrics) + "\n"
         return out
 
-    def _summarize_iter(self) -> str:
-        out = ""
-        if (
-            self.iter_total_s.get()
-            or self.iter_wait_s.get()
-            or self.iter_next_batch_s.get()
-            or self.iter_format_batch_s.get()
-            or self.iter_get_s.get()
-        ):
-            out += "\nDataset iterator time breakdown:\n"
-            out += "* In ray.wait(): {}\n".format(fmt(self.iter_wait_s.get()))
-            out += "* In ray.get(): {}\n".format(fmt(self.iter_get_s.get()))
-            out += "* In next_batch(): {}\n".format(fmt(self.iter_next_batch_s.get()))
-            out += "* In format_batch(): {}\n".format(
-                fmt(self.iter_format_batch_s.get())
-            )
-            out += "* In user code: {}\n".format(fmt(self.iter_user_s.get()))
-            out += "* Total time: {}\n".format(fmt(self.iter_total_s.get()))
-        return out
+    def get_total_wall_time(self) -> float:
+        parent_sum = sum(p.get_total_wall_time() for p in self.parents)
+        return parent_sum + sum(
+            getattr(ss, StageStatsSummary.WALL_TIME).get("max", 0)
+            for ss in self.stages_stats
+        )
+
+    def get_total_cpu_time(self) -> float:
+        parent_sum = sum(p.get_total_cpu_time() for p in self.parents)
+        return parent_sum + sum(
+            getattr(ss, StageStatsSummary.CPU_TIME).get("sum", 0)
+            for ss in self.stages_stats
+        )
+
+    def get_max_heap_memory(self) -> float:
+        parent_memory = [p.get_max_heap_memory() for p in self.parents]
+        parent_max = max(parent_memory) if parent_memory else 0
+        return max(
+            parent_max,
+            *[
+                getattr(ss, StageStatsSummary.PEAK_HEAP_MEMORY).get("max", 0)
+                for ss in self.stages_stats
+            ],
+        )
+
+
+@dataclass
+class StageStatsSummary:
+    # Keys used in initialization params
+    BLOCK_EXECUTION_SUMMARY = "block_execution_summary_str"
+    WALL_TIME = "wall_time"
+    CPU_TIME = "cpu_time"
+    PEAK_HEAP_MEMORY = "memory"
+    OUTPUT_NUM_ROWS = "output_num_rows"
+    OUTPUT_SIZE_BYTES = "output_size_bytes"
+    NODE_COUNT = "node_count"
+
+    stage_name: str
+    is_substage: bool
+    time_total_s: float
+    block_execution_summary_str: str
+    wall_time: Optional[Dict[str, float]] = None
+    cpu_time: Optional[Dict[str, float]] = None
+    memory: Optional[Dict[str, float]] = None
+    output_num_rows: Optional[Dict[str, float]] = None
+    output_size_bytes: Optional[Dict[str, float]] = None
+    node_count: Optional[Dict[str, float]] = None
+
+    @classmethod
+    def from_block_metadata(
+        cls,
+        block_metas: List[BlockMetadata],
+        time_total_s: float,
```

Comment:
Is this the walltime of the stage? If so, what's the difference between this and computed walltime?

Reply:
The parameter passed as `time_total_s` is from `DatasetStats.time_total_s`, pulled from the stats actor. 
`stats_map, self.time_total_s = ray.get(ac.get.remote(self.stats_uuid))`
This is the total walltime of the overall stage, whereas the computed walltimes from `from_block_metadata` could be on a substage level. The `DatasetStats.time_total_s` is needed to populate the execution summary string present in the original `summary_string()` method: https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/stats.py#L250-L280

Reply:
Since we are going to deprecate stats actor, how is `time_total_s` going to be maintain in the future?
The max() across all bocks may not be the same as `time_total_s`, as they may not start the execution simultaneously.

Reply:
Currently, the stats actor `get()` method calculates `time_total_s` as the duration between the task end/start times: https://github.com/ray-project/ray/blob/f491e5c4a35390c6eb1a3147b030c90965bc9b57/python/ray/data/_internal/stats.py#L138
https://github.com/ray-project/ray/blob/f491e5c4a35390c6eb1a3147b030c90965bc9b57/python/ray/data/_internal/lazy_block_list.py#L687

This PR doesn't change the logic of the `time_total_s` calculation from the current implementation. In the future, we would need to move this logic of recording start/end times into the `BlockMetadata` class, so that we can do this calculation without the stats actor.